### PR TITLE
feat: add sudo-prompt to start on linux

### PR DESCRIPTION
**Pull request by Shou**
We prompt the user with a sudo-prompt upon launch on Linux-based systems
to ensure the program has enough permissions for features needed
throughout the program's runtime.

Changelog-Entry: Add a sudo-prompt upon launch on Linux-based systems.

### CODE REVIEW

**Comment by Shou** [comment]
Here's the polkit policy file I'm using. `/usr/share/polkit-1/actions/etcher.policy`

```xml
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE policyconfig PUBLIC
 "-//freedesktop//DTD PolicyKit Policy Configuration 1.0//EN"
 "http://www.freedesktop.org/standards/PolicyKit/1/policyconfig.dtd">
<policyconfig>

  <vendor>Examples for the PolicyKit Project</vendor>
  <vendor_url>http://hal.freedesktop.org/docs/PolicyKit/</vendor_url>

  <action id="org.freedesktop.policykit.pkexec.etcher">
    <description>Run Etcher</description>
    <message>Authentication is required to run Etcher</message>
    <icon_name>audio-x-generic</icon_name> 
    <defaults>
      <allow_any>auth_admin</allow_any>
      <allow_inactive>auth_admin</allow_inactive>
      <allow_active>auth_admin</allow_active>
    </defaults>
    <annotate key="org.freedesktop.policykit.exec.path">/home/shou/Prog/Javascript/etcher-alt/node_modules/electron/dist/electron</annotate>
    <annotate key="org.freedesktop.policykit.exec.allow_gui">true</annotate>
  </action>

</policyconfig>
```

**Comment by Shou** [comment]
Here we go, AppImages should work – thoughts on these changes?

File: lib/start.js [hunk 1]
```diff
@@ -29,5 +29,69 @@
...
+        name: 'gksudo',
+        options: [
+          '--preserve-env',
+          '--message',
+          'Please provide your password. Etcher needs special permissions to interact with raw devices.'
+        ]
+      },
+      {
+        name: 'kdesudo',
+        options: []
```
Comment (jviotti) [inline lib/start.js]: Can we customize `kdesudo`'s window message like we do with `gksu`?

File: lib/start.js [hunk 2]
```diff
@@ -29,5 +29,69 @@
...
+          throw new Error(`Command does not exist: ${command.name}`)
+        }
+
+        return command
+      })
+    })).then((command) => {
+      return new Bluebird((resolve, reject) => {
+        const args = process.env.APPIMAGE ? [ process.env.APPIMAGE ] : process.argv
+        const appinstallSkip = !process.env.APPIMAGE ? {} : {
+          SKIP: 1
```
Comment (jviotti) [inline lib/start.js]: We could always pass this variable (since its harmless) and reduce a bit the complexity

File: lib/child-writer/writer-proxy.js [hunk 3]
```diff
@@ -92,6 +92,7 @@ permissions.isElevated().then((elevated) => {
         ELECTRON_RUN_AS_NODE: 1,
         IPC_SERVER_ID: process.env.IPC_SERVER_ID,
         IPC_CLIENT_ID: process.env.IPC_CLIENT_ID,
+        ETCHER_RUNNING: 1,
```
Comment (jviotti) [inline lib/child-writer/writer-proxy.js]: What is this variable for? What does it mean?

File: electron-builder.yml [hunk 4]
```diff
@@ -83,6 +83,7 @@ deb:
     - libxss1
     - libxtst6
     - polkit-1-auth-agent | policykit-1-gnome | polkit-kde-1
+    - gksu | kdesudo
```
Comment (jviotti) [inline electron-builder.yml]: Should we add the rpm equivalents below?

File: lib/child-writer/writer-proxy.js (hunk 3 above)
Comment (Shou) [reply lib/child-writer/writer-proxy.js]: Basically, we need some way to communicate to the writer-proxy that it was called. Previously it was just a check of whether it was running with privileges or not, but now it always does. Knowing this, it creates the IPC channel. We could use `SKIP` instead but I thought it'd be easier to understand with a new variable. I'll add a comment explaining this.

Comment (jviotti) [reply lib/child-writer/writer-proxy.js]: So basically if you elevate the app with `gksudo`, then `is-elevated` still returns `false`?

Comment (Shou) [reply lib/child-writer/writer-proxy.js]: Oh no, the opposite, it always returns true which causes it to skip the step and no IPC channel is established which means no progress.

**Comment by Shou** [comment]
Testing `beesu` on Fedora, the `--preserve-environment` flag doesn't work, nor the shorthand `-m`, which presents the AppImage installation dialogue a second time, otherwise seems to work.

**Comment by Shou** [comment]
This is done and ready for reviews, the CI tests have been stuck for a while.

**Comment by jhermsmeier** [comment]
@Shou can you commit --amend and push to this again to re-trigger the CI builds? AppVeyor should be unstuck now.

File: lib/child-writer/writer-proxy.js (hunk 3 above)
Comment (jhermsmeier) [reply lib/child-writer/writer-proxy.js]: @Shou have you added the comment explaining this yet? I'm not seeing it, or I'm missing something.

File: lib/start.js [hunk 5]
```diff
@@ -29,5 +29,110 @@
...
+  const os = require('os')
+  const _ = require('lodash')
+  const Bluebird = require('bluebird')
+  const commandExists = Bluebird.promisify(require('command-exists'))
+  const childProcess = require('child_process')
+  const permissions = require('./shared/permissions')
+  const errors = require('./shared/errors')
+  const EXIT_CODES = require('./shared/exit-codes')
+
+  const MESSAGE = 'Etcher needs special permissions to interact with raw devices. Please provide your password.'
```
Comment (jhermsmeier) [inline lib/start.js]: Should this live in `shared/messages` instead?

File: electron-builder.yml [hunk 6]
```diff
@@ -83,10 +83,12 @@ deb:
...
 rpm:
   icon: assets/icon.png
   depends:
     - lsb
     - libXScrnSaver
+    - beesu
```
Comment (jviotti) [inline electron-builder.yml]: So `kdesudo` is not available on Fedora?

File: lib/child-writer/renderer-utils.js [hunk 7]
```diff
@@ -46,5 +46,5 @@ exports.getApplicationEntryPoint = () => {
   // On GNU/Linux, `pkexec` resolves relative paths
   // from `/root`, therefore we pass an absolute path,
   // in order to be on the safe side.
-  return path.join(CONSTANTS.PROJECT_ROOT, relativeEntryPoint)
+  return path.resolve(CONSTANTS.PROJECT_ROOT, relativeEntryPoint)
```
Comment (jviotti) [inline lib/child-writer/renderer-utils.js]: Can you explain why this change is needed? Will it break other platforms? If not, should we send it as a different PR?

**Comment by jviotti** [comment]
I'll take over this PR so we can release today.

File: lib/child-writer/renderer-utils.js [hunk 8]
```diff
@@ -43,8 +43,5 @@ exports.getApplicationEntryPoint = () => {
   const ENTRY_POINT_ARGV_INDEX = 1
   const relativeEntryPoint = electron.remote.process.argv[ENTRY_POINT_ARGV_INDEX]
 
-  // On GNU/Linux, `pkexec` resolves relative paths
-  // from `/root`, therefore we pass an absolute path,
-  // in order to be on the safe side.
```
Comment (lurch) [inline lib/child-writer/renderer-utils.js]: I've not been following the progress of this PR, but why has this comment been removed?

**Comment by jviotti** [comment]
I tested this on Ubuntu (gtksu and kdesudo in AppImage and deb) and Fedora (beesu on rpm) and works fine. 

@Shou Can you add a bit TODO comment about the `ETCHER_RUNNING` environment variable, and potentially create an issue for it once this gets merged? I'd love to take a closer look later on.

File: lib/child-writer/renderer-utils.js (hunk 8 above)
Comment (Shou) [reply lib/child-writer/renderer-utils.js]: We don't use `pkexec` anymore so this explanation isn't necessary

**Comment by jhermsmeier** [comment]
@Shou how're we looking on this? There are still some requested changes & the branch also needs some rebasing / conflict resolution.

**Comment by jviotti** [comment]
@Shou @jhermsmeier @lurch I can confirm that removing the `ETCHER_RUNNING` has no effect. The flash process works fine without it on my Ubuntu 16.04 system, so I think we can safely get rid of it.
